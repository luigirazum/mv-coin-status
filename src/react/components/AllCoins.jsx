import { useSelector, shallowEqual } from 'react-redux';
import { Link } from 'react-router-dom';
import Coin from './Coin';

const AllCoins = () => {
  const coins = useSelector((store) => store.coins.assets, shallowEqual);

  return (
    <section className="allCoins">
      {coins.map((coin) => (
        <Link to={coin.id} key={coin.id} className="coinLink">
          <Coin id={coin.id} />
        </Link>
      ))}
    </section>
  );
};

export default AllCoins;
