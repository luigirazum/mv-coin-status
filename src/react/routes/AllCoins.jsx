import { useSelector } from 'react-redux';

const AllCoins = () => {
  const coins = useSelector((store) => store.coins.assets);

  return (
    <section>
      <h2>All Coins</h2>
      <ul>
        {coins.map((coin) => (
          <li key={coin.id}>
            {coin.name}
            <p>{coin.rank}</p>
            <p>{coin.symbol}</p>
          </li>
        ))}
      </ul>
    </section>
  );
};

export default AllCoins;
