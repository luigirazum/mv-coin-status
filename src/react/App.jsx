import { Routes, Route } from 'react-router-dom';
import Home from './routes/Home';
import CoinsLayout from './routes/CoinsLayout';
import AllCoins from './routes/AllCoins';
import CoinDetails from './routes/CoinDetails';
import '../styles/App.css';

const App = () => (
  <Routes>
    <Route index element={<Home />} />
    <Route path="coins" caseSensitive element={<CoinsLayout />}>
      <Route index caseSensitive element={<AllCoins />} />
      <Route path=":id" caseSensitive element={<CoinDetails />} />
    </Route>
    <Route path="*" element={<h3>ERROR</h3>} />
  </Routes>
);

export default App;
