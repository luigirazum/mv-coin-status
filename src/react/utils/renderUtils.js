import PropTypes from 'prop-types';
import React from 'react';
import {
  BrowserRouter, MemoryRouter, Routes, Route,
} from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// import your setupStore as a basic setup
import setupStore from '../../redux/store';

/** ------------------------------------------------------------
 *   > to test components that require a store/state/router
 *  ------------------------------------------------------------ */
const renderWithProvider = (
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = {},
  {
    routerWrapper = null,
    routes = ['/'],
    path = '/',
    user = userEvent,
  } = {},
) => {
  const RouterWrapper = ({ children }) => {
    if (routerWrapper.memory) {
      return (
        <MemoryRouter initialEntries={routes}>
          <Routes>
            <Route path={path} element={children} />
          </Routes>
        </MemoryRouter>
      );
    }

    return (
      <BrowserRouter>
        <Routes>
          <Route path={path} element={children} />
        </Routes>
      </BrowserRouter>
    );
  };

  RouterWrapper.propTypes = {
    children: PropTypes.element.isRequired,
  };

  const Wrapper = ({ children }) => (
    <Provider store={store}>
      {
      routerWrapper
        ? <RouterWrapper>{children}</RouterWrapper>
        : children
      }
    </Provider>
  );

  Wrapper.propTypes = {
    children: PropTypes.element.isRequired,
  };

  const component = Wrapper({ children: ui });
  if (routerWrapper && routerWrapper.browser) window.history.pushState({}, 'test page', routes[0]);

  // Return an object with the store and all of RTL's query functions
  return ({
    store,
    component,
    user,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  });
};

export default renderWithProvider;
