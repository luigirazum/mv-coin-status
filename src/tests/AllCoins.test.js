import { screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { act } from 'react-test-renderer';
import { renderWithProviders } from '../react/utils/renderUtils';
import { mockCoins } from './mocks/mockData';
import AllCoins from '../react/routes/AllCoins';

let testRenderWithProviders = null;
let store = null;
let container = null;
/**
 * START: AllCoins component test
*/
describe('AllCoins component tests', () => {
  /**
   * START: AllCoins render
  */
  describe('Coins must be rendered', () => {
    beforeEach(() => {
      // setup a DOM element as a render target
      testRenderWithProviders = renderWithProviders(
        <MemoryRouter initialEntries={['/coins']}>
          <AllCoins />
        </MemoryRouter>,
        {
          preloadedState: {
            coins: {
              assets: [...mockCoins.coins],
              status: { type: 'idle' },
              filter: { active: false, by: '' },
              error: null,
            },
          },
        },
      );
      ({ container } = testRenderWithProviders);
      document.body.appendChild(container);
    });

    afterEach(() => {
      //  cleanup on exiting
      // unmountComponentAtNode(container);
      container.remove();
      container = null;
      testRenderWithProviders = null;
    });

    test('should render bitcoin', () => {
      expect(screen.getByTestId('bitcoin')).toBeInTheDocument();
    });
    test('should render ethereum', () => {
      expect(screen.getByTestId('ethereum')).toBeInTheDocument();
    });
    test('should render tether', () => {
      expect(screen.getByTestId('tether')).toBeInTheDocument();
    });
    test('should render binance-coin', () => {
      expect(screen.getByTestId('binance-coin')).toBeInTheDocument();
    });
    test('should render usd-coin', () => {
      expect(screen.getByTestId('usd-coin')).toBeInTheDocument();
    });
    test('should render 5 coins in total', () => {
      expect(screen.getAllByTestId('coinName')).toHaveLength(5);
    });
  });
  /**
   * END: AllCoins render
  */

  describe('AllCoins component filter test', () => {
    beforeEach(() => {
      testRenderWithProviders = renderWithProviders(
        <MemoryRouter initialEntries={['/coins']}>
          <AllCoins />
        </MemoryRouter>,
        {
          preloadedState: {
            coins: {
              assets: [...mockCoins.coins],
              status: { type: 'idle' },
              filter: { active: false, by: '' },
              error: null,
            },
          },
        },
      );
      ({ container, store } = testRenderWithProviders);
    });

    afterEach(() => {
      container.remove();
      container = null;
      store = null;
      testRenderWithProviders = null;
    });

    test('no filter should render 5 coins', () => {
      expect(screen.getAllByTestId('coinName')).toHaveLength(5);
    });

    test('with filter -coin- should render 2 coins', () => {
      act(() => {
        store.dispatch({ type: 'coins/filterCoins', payload: 'coin' });
      });
      expect(screen.getAllByTestId('coinName')).toHaveLength(2);
    });

    test('with filter -ether- should render 2 coins', () => {
      act(() => {
        store.dispatch({ type: 'coins/filterCoins', payload: 'ether' });
      });
      expect(screen.getAllByTestId('coinName')).toHaveLength(2);
    });

    test('with filter -allcoins- should render 0 coins', () => {
      act(() => {
        store.dispatch({ type: 'coins/filterCoins', payload: 'allcoins' });
      });
      expect(screen.queryByTestId('coinName')).toBeNull();
    });
  });

  describe('AllCoins component snapshot test', () => {
    test('snapshot renders 5 Coins correctly', () => {
      container = renderWithProviders(
        <MemoryRouter initialEntries={['/coins']}>
          <AllCoins />
        </MemoryRouter>,
        {
          preloadedState: {
            coins: {
              assets: [...mockCoins.coins],
              status: { type: 'idle' },
              filter: { active: false, by: '' },
              error: null,
            },
          },
        },
      ).container;

      expect(container).toMatchSnapshot();
    });
  });
});
/**
 * END: AllCoins component tests
*/
