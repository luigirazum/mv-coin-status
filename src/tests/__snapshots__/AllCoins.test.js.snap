// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllCoins component tests AllCoins component snapshot test snapshot renders 5 Coins correctly 1`] = `
<section
  className="allCoins"
>
  <a
    className="coinLink evenCoin"
    href="/coins/bitcoin"
    onClick={[Function]}
  >
    <article
      className="coinItem"
      data-testid="bitcoin"
    >
      <section
        className="itemSection itemHeader"
      >
        <h3
          className="itemName"
          data-testid="coinName"
        >
          Bitcoin
        </h3>
        <div
          className="iconContainer"
        >
          <img
            alt="Bitcoin"
            className="coinIcon"
            data-testid="coinLogo"
            src="https://static.coinstats.app/coins/1650455588819.png"
          />
        </div>
      </section>
      <section
        className="itemSection"
      >
        <div
          className="itemDescription"
        >
          <h3
            className="itemCategory rankCategory"
          >
            Rank
          </h3>
          <p
            className="itemData rankData"
            data-testid="coinRank"
          >
            1
          </p>
        </div>
        <div
          className="itemDescription"
        >
          <h4
            className="itemCategory"
          >
            Symbol
          </h4>
          <p
            className="itemData"
            data-testid="coinSymbol"
          >
            BTC
          </p>
        </div>
      </section>
    </article>
  </a>
  <a
    className="coinLink oddCoin"
    href="/coins/ethereum"
    onClick={[Function]}
  >
    <article
      className="coinItem"
      data-testid="ethereum"
    >
      <section
        className="itemSection itemHeader"
      >
        <h3
          className="itemName"
          data-testid="coinName"
        >
          Ethereum
        </h3>
        <div
          className="iconContainer"
        >
          <img
            alt="Ethereum"
            className="coinIcon"
            data-testid="coinLogo"
            src="https://static.coinstats.app/coins/1650455629727.png"
          />
        </div>
      </section>
      <section
        className="itemSection"
      >
        <div
          className="itemDescription"
        >
          <h3
            className="itemCategory rankCategory"
          >
            Rank
          </h3>
          <p
            className="itemData rankData"
            data-testid="coinRank"
          >
            2
          </p>
        </div>
        <div
          className="itemDescription"
        >
          <h4
            className="itemCategory"
          >
            Symbol
          </h4>
          <p
            className="itemData"
            data-testid="coinSymbol"
          >
            ETH
          </p>
        </div>
      </section>
    </article>
  </a>
  <a
    className="coinLink evenCoin"
    href="/coins/tether"
    onClick={[Function]}
  >
    <article
      className="coinItem"
      data-testid="tether"
    >
      <section
        className="itemSection itemHeader"
      >
        <h3
          className="itemName"
          data-testid="coinName"
        >
          Tether
        </h3>
        <div
          className="iconContainer"
        >
          <img
            alt="Tether"
            className="coinIcon"
            data-testid="coinLogo"
            src="https://static.coinstats.app/coins/1650455771843.png"
          />
        </div>
      </section>
      <section
        className="itemSection"
      >
        <div
          className="itemDescription"
        >
          <h3
            className="itemCategory rankCategory"
          >
            Rank
          </h3>
          <p
            className="itemData rankData"
            data-testid="coinRank"
          >
            3
          </p>
        </div>
        <div
          className="itemDescription"
        >
          <h4
            className="itemCategory"
          >
            Symbol
          </h4>
          <p
            className="itemData"
            data-testid="coinSymbol"
          >
            USDT
          </p>
        </div>
      </section>
    </article>
  </a>
  <a
    className="coinLink oddCoin"
    href="/coins/binance-coin"
    onClick={[Function]}
  >
    <article
      className="coinItem"
      data-testid="binance-coin"
    >
      <section
        className="itemSection itemHeader"
      >
        <h3
          className="itemName"
          data-testid="coinName"
        >
          BNB
        </h3>
        <div
          className="iconContainer"
        >
          <img
            alt="BNB"
            className="coinIcon"
            data-testid="coinLogo"
            src="https://static.coinstats.app/coins/1666608145347.png"
          />
        </div>
      </section>
      <section
        className="itemSection"
      >
        <div
          className="itemDescription"
        >
          <h3
            className="itemCategory rankCategory"
          >
            Rank
          </h3>
          <p
            className="itemData rankData"
            data-testid="coinRank"
          >
            4
          </p>
        </div>
        <div
          className="itemDescription"
        >
          <h4
            className="itemCategory"
          >
            Symbol
          </h4>
          <p
            className="itemData"
            data-testid="coinSymbol"
          >
            BNB
          </p>
        </div>
      </section>
    </article>
  </a>
  <a
    className="coinLink evenCoin"
    href="/coins/usd-coin"
    onClick={[Function]}
  >
    <article
      className="coinItem"
      data-testid="usd-coin"
    >
      <section
        className="itemSection itemHeader"
      >
        <h3
          className="itemName"
          data-testid="coinName"
        >
          USD Coin
        </h3>
        <div
          className="iconContainer"
        >
          <img
            alt="USD Coin"
            className="coinIcon"
            data-testid="coinLogo"
            src="https://static.coinstats.app/coins/1650455825065.png"
          />
        </div>
      </section>
      <section
        className="itemSection"
      >
        <div
          className="itemDescription"
        >
          <h3
            className="itemCategory rankCategory"
          >
            Rank
          </h3>
          <p
            className="itemData rankData"
            data-testid="coinRank"
          >
            5
          </p>
        </div>
        <div
          className="itemDescription"
        >
          <h4
            className="itemCategory"
          >
            Symbol
          </h4>
          <p
            className="itemData"
            data-testid="coinSymbol"
          >
            USDC
          </p>
        </div>
      </section>
    </article>
  </a>
</section>
`;
